src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
+                    nsae);
+            }
+        } else if (derivationParameterSpec instanceof HKDFParameterSpec.ExtractThenExpand) {
+            HKDFParameterSpec.ExtractThenExpand anExtractThenExpand =
+                (HKDFParameterSpec.ExtractThenExpand) derivationParameterSpec;
+            ikms = anExtractThenExpand.ikms();
+            salts = anExtractThenExpand.salts();
+            // we should be able to combine these Lists of keys into single
+            // SecretKey Objects,
+            // unless we were passed something bogus or an unexportable P11 key
+            try {
+                inputKeyMaterial = consolidateKeyMaterial(ikms);
+                salt = consolidateKeyMaterial(salts);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "Issue encountered when combining ikm or salt values into"
+                    + " single keys").initCause(ike);
+            }
+            // set this value in the "if"
+            if ((info = anExtractThenExpand.info()) == null) {
+                info = new byte[0];
+            }
+            length = anExtractThenExpand.length();
+            if (length > (hmacLen * 255)) {
+                throw new InvalidAlgorithmParameterException(
+                    "Requested length exceeds maximum allowed key stream "
+                    + "length");
+            }
+            // perform extract and then expand
+            try {
+                byte[] extractResult = hkdfExtract(inputKeyMaterial, (salt
+                                                                      == null) ? null : salt.getEncoded());
+                pseudoRandomKey = new SecretKeySpec(extractResult, "RAW");
+                return Arrays.copyOf(hkdfExpand(pseudoRandomKey, info, length),
+                                     length);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "an HKDF ExtractThenExpand could not be initialized with "
+                    + "the given key or salt material").initCause(ike);
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
+                    nsae);
+            }
+        }
+        throw new InvalidAlgorithmParameterException(
+            "an HKDF could not be initialized with the given KDFParameterSpec");
```

Comment:
It's clearer to state that the given `KDFParameterSpec` must be `HKDFParameterSpec`. Also, given that KDF.getInstance() takes a KDFParameters parameters, I'd avoid the word "initialized" as it may confuse people which parameters you are talking about.
I'd suggest something like "HKDF data/key derivatopn requires HKDFParameterSpec, not " + derivationParameterSpec.getClass()
Also, for readability, it may be better to check the specified `derivationParameterSpec` is an instanceof `HKDFParameterSpec` in the beginning.

Reply:
Yes, resolved.